Proposal: add Clearing the associations table

infograf768 opened the pull request:
In some cases, one would like to clear all the records from the associations table.
This PR adds a Toolbar button in the list view to do so.
The button will display only when no Item Type is selected. It displays an alert when clicked.

![screen shot 2016-08-12 at 09 43 22](https://cloud.githubusercontent.com/assets/869724/17615864/48e90050-6071-11e6-8e34-b28ec2883f40.png)

![screen shot 2016-08-12 at 09 44 09](https://cloud.githubusercontent.com/assets/869724/17615875/5dfc727e-6071-11e6-8176-572ca58f7772.png)

![screen shot 2016-08-12 at 09 44 44](https://cloud.githubusercontent.com/assets/869724/17615881/7069e9aa-6071-11e6-85b9-cfc6433145a2.png)

infograf768 committed:
Clear the associations table

andrepereiradasilva commented on administrator/components/com_associations/language/en-GB/en-GB.com_associations.ini:
```diff
@@ -29,6 +29,11 @@ COM_ASSOCIATIONS_HEADING_MENUTYPE_ASC="Menu ascending"
+COM_ASSOCIATION_PURGE_CONFIRM_PROMPT="Are you sure you want to clear all associations? Confirming will permanently delete them!"
+; Do not translate the name of the table
+COM_ASSOCIATIONS_PURGE_FAILED="Failed to clear the 'associations' table."
+COM_ASSOCIATIONS_PURGE_SUCCESS="The 'associations' table was successfully cleared."
```
Proposal:
COM_ASSOCIATIONS_PURGE="Delete all associations"
COM_ASSOCIATION_PURGE_CONFIRM_PROMPT="Are you sure you want to delete all associations? Confirming will permanently delete them!"
COM_ASSOCIATIONS_PURGE_FAILED="Failed to delete all the associations."
COM_ASSOCIATIONS_PURGE_SUCCESS="All associations have been deleted."

infograf768 replied:
Core uses "clear"

andrepereiradasilva replied:
![image](https://cloud.githubusercontent.com/assets/9630530/17617046/a7b74f30-606f-11e6-8b40-77d0cc4d7c5b.png)

infograf768 replied:
That is because these are files.

Look rather at:
![screen shot 2016-08-12 at 10 38 53](https://cloud.githubusercontent.com/assets/869724/17617197/0fb025cc-6079-11e6-9b42-1db9c8d69203.png)

and for finder Index
![screen shot 2016-08-12 at 10 40 41](https://cloud.githubusercontent.com/assets/869724/17617241/4dce0216-6079-11e6-9f8b-a969980e7e93.png)

infograf768 replied:
I do agree though that it is not consequent in core

![screen shot 2016-08-12 at 10 43 44](https://cloud.githubusercontent.com/assets/869724/17617313/bc56dac8-6079-11e6-8543-bd36d85cda1d.png)

andrepereiradasilva replied:
the core yet again does not make sense
the trash icon is for trashing ie changing it's state to trash in the db
the delete icon is for deleting ie deleting from db

andrepereiradasilva replied:
and BTW the update will be changed in https://github.com/joomla/joomla-cms/pull/11423
so the only bad one is the finder

infograf768 commented:
I am now working also on a Rebuild table button to get rid of possible orphans.

andrepereiradasilva commented on administrator/components/com_associations/models/associations.php:
```diff
@@ -348,4 +348,35 @@ protected function getListQuery()
+	 *
+	 * @return  bool True on success
+	 *
+	 * @throws	Exception
```
it doesn't trow an exception.

andrepereiradasilva commented on administrator/components/com_associations/models/associations.php:
```diff
@@ -348,4 +348,35 @@ protected function getListQuery()
+		$db = $this->getDbo();
+
+		// Get the localise data
+		$db->setQuery('TRUNCATE TABLE #__associations');
```
Proposal

``` php
// Delete all table data.
$db->setQuery('TRUNCATE TABLE ' . $db->quoteName('#__associations'));
```

andrepereiradasilva commented on administrator/components/com_associations/models/associations.php:
```diff
@@ -348,4 +348,35 @@ protected function getListQuery()
+		}
+		catch (JDatabaseExceptionExecuting $e)
+		{
+			$this->_message = JText::_('COM_ASSOCIATIONS_PURGE_FAILED');
```
``` php
JFactory::getApplication()->enqueueMessage(JText::_('COM_ASSOCIATIONS_PURGE_FAILED', 'error'));
```

alikon commented:
I like it

infograf768 committed:
correcting various suggestions by andre. Thanks.

alikon commented:
i like this feature

infograf768 committed:
Correcting some errors. Implementing orphans deleting.

alikon commented:
i've tested successfully 

RFC
in this way we can only "delete all" associated items
what about being able to  be more granular  i.e. delete all:
- associated articles
-  associated articles categories
-  associated contacts
-  associated contacts catgegories
-  etc

p.s.
tested before the addition of orphans deleting

infograf768 commented:
Corrected lang strings (to please @andrepereiradasilva  😄 )

Implemented cleaning orphans

![screen shot 2016-08-13 at 08 15 02](https://cloud.githubusercontent.com/assets/869724/17641787/37438ef2-612e-11e6-9f71-4313dce45add.png)

To test orphans deleting, associate 2 items.
With PhpMyAdmin, delete one of them in the associations table so that an orphan remains
For example you would get here an orphaned menu item association:

![screen shot 2016-08-12 at 11 33 40](https://cloud.githubusercontent.com/assets/869724/17641793/6ee01010-612e-11e6-8c6f-7c5d5eb8a2f0.png)

Click on the button.

infograf768 commented:
Concerning more granular delete.
Not so easy as the associations table does not differentiate for example between categories, languages, etc., only context which corresponds to $associationsContext
Therefore much more complex to code

alikon commented:
we should change the icon of one of the 2 buttons they are the same , maybe can be confusing

infograf768 committed:
cs

infograf768 commented:
btw, I am conscious that the clean Orphans method can be optimized by using simplified queries as I use 3. @alikon

infograf768 commented:
For the Icon, we could use for one of them the Empty Trash icon (i.e. delete)

alikon commented on administrator/components/com_associations/views/associations/view.html.php:
```diff
@@ -181,6 +181,11 @@ protected function addToolbar()
+			if (!isset($this->component))
+			{
+				JToolbarHelper::custom('associations.purge', 'purge', 'purge', 'COM_ASSOCIATIONS_PURGE', false, false);
+				JToolbarHelper::custom('associations.clean', 'purge', 'purge', 'COM_ASSOCIATIONS_DELETE_ORPHANS', false, false);
```
can we use refresh instead of purge for orphans button

alikon commented:
@infograf768  not so sure if we can simplify using 1 complex query instead of the 3 as it is now, but i'll look at it

alikon commented:
testing the orphans feature:
if i don't have orphans  should tell me something like nothing to delete/no orphans found
instead of the standard message
![orphans](https://cloud.githubusercontent.com/assets/181681/17641920/878b922e-6133-11e6-9bf7-66010ca4cab1.PNG)

infograf768 commented:
> if i don't have orphans should tell me something like nothing to delete/no orphans found
> instead of the standard message

Good idea.

alikon committed:
from 3 to 2

one query less

infograf768 committed:
Changing icon. Adding message if no orphan to delete.

infograf768 commented:
@alikon 
Thanks for the simplified query.

Changed icon, added message when no orphans to delete.
![screen shot 2016-08-13 at 10 17 52](https://cloud.githubusercontent.com/assets/869724/17642285/8d54072a-613f-11e6-96d1-fcc685b1cdf0.png)

Please test again.

alikon commented:
ok  for me now
just still some concern about delete orphans performance 
if  we need a no-unique index  on key field ..... i need further investigation on this 

and for a proper message if there is nothing to de-associate/clear all associations

but these apart,,,:thumbsup:

infograf768 committed:
Adding message when Deleting all associations and the associations table
is empty

infograf768 commented:
Added message when Deleting all associations and the table was empty:
`COM_ASSOCIATIONS_PURGE_NONE="There were no associations to delete."`

andrepereiradasilva committed:
revert spaces

andrepereiradasilva committed:
revert spaces 2

andrepereiradasilva committed:
revert spaces 3

andrepereiradasilva committed:
now it's called itemType

andrepereiradasilva commented:
merging as decide. will add ACL checks in the controller later
